platform jitter entropy

lws-team opened the pull request:
At the moment if no HWRNG support as on Hikey, a deterministic RC4 PRNG is used.  That's not nice for the reasons listed in the first patch.

These two patches add an internal api that indicates it is a good time to harvest jitter from a platform time source.  By default it does nothing with an empty weak implementation.

An implementation for Hikey is also added using the ARMv8 TSC.

The api to harvest entropy is added to the PRNG init, where it gets 16 bits in the hikey implementation, and at the start of RPC calls, where it harvests 2 bits of jitter.

jforissier commented on core/arch/arm/plat-hikey/main.c:
```diff
@@ -205,4 +206,47 @@ static TEE_Result peripherals_init(void)
+ * and add entropy when we reach 8 bits.
+ */
+
+void plat_prng_add_jitter_entropy(void)
```
That's more or less what `inject_entropy_with_timestamp()` does (`entry_std.c`), no? Either one or the other but we don't want two ways of adding time-based entropy.

jforissier commented on core/arch/arm/plat-hikey/main.c:
```diff
@@ -205,4 +206,47 @@ static TEE_Result peripherals_init(void)
+		__asm__("mrc p15, 0, %0, c9, c13, 0" : "=r"(tsc32));
+		tsc = tsc32;
+	}
+#endif
```
Avoid inline assembly in `.c` files. See `core/arch/arm/include/arm{,32,64}.h`.
As I said above, why would this be better than using `tee_time_get_sys_time()` as a source of entropy? `tee_time_get_sys_time()` is portable, and is typically based on CNTPCT/CNTPCT_EL0.

jforissier commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1193,6 +1194,8 @@ static uint32_t rpc_cmd_nolock(uint32_t cmd, size_t num_params,
 
 	assert(arg && carg && num_params <= THREAD_RPC_MAX_NUM_PARAMS);
 
+	plat_prng_add_jitter_entropy();
```
+1 for gathering entropy here, indeed.

lws-team commented:
I didn't see there was already some attempt to do this... however

 - we are talking about a very different scale of jitter.  This patch samples a counter that operates at the core clock rate.  So things like missed branches, drift in async clocks also used by peripherals to complete tasks, how busy the DDR or busses are etc appear there especially in the LSBs which reflect sub-ns differences in performance.  This is much more finegrained than microsecond resolution in the sys time.

 - what you are doing atm

```
static void inject_entropy_with_timestamp(void)
{
        TEE_Time current;

        if (tee_time_get_sys_time(&current) == TEE_SUCCESS)
                tee_prng_add_entropy((uint8_t *)&current, sizeof(current));
}
```

injects a lot of non-entropy... the whole of the seconds part that is in the TEE_Time struct is predictable as is much of the microseconds part.  We're supposed to be dripping a little bit of high quality random into the PRNG at intervals, not spamming 00 00 00 SS etc, which fills the PRNG entropy buffer and causes it to ignore further entropy additions.

jforissier commented:
OK, so if the current code is bad (which I'm willing to admit), then we need to throw it away at the same time we add your new code. Can you please add a patch that would do just that?

lws-team commented:
OK... thanks... as I say I missed this existing implementation until you pointed it out.  So I will study it and update later.

BTW Intel has the same timestamp counter concept and an instruction "rdtsc".

jforissier commented:
Good. Thanks for sharing your work, that's definitely helpful.

lws-team commented:
This update should cover everything discussed... plus or minus whatever Travis finds.  In addition to the points raised:

I moved the implementation to core/arch/arm, so it covers all arm devices.

I also made the weak default the old system time implementation for backwards compatibility for those without a platform / arch implementation.

jforissier commented on core/arch/arm/kernel/tee_time_arm_cntpct.c:
```diff
@@ -54,3 +56,42 @@ static const struct time_source arm_cntpct_time_source = {
+
+void plat_prng_add_jitter_entropy(void)
+{
+	uint64_t tsc = read_cntvct_el0();
```
Why not use CNTPCT? (this file is named `tee_time_arm_cntpct.c`)

jforissier commented on core/tee/tee_cryp_utl.c:
```diff
@@ -378,6 +378,16 @@ TEE_Result tee_prng_add_entropy(const uint8_t *in, size_t len)
+}
+
+
+
```
One blank line is enough

jforissier commented on core/arch/arm/kernel/tee_time_arm_cntpct.c:
```diff
@@ -54,3 +56,42 @@ static const struct time_source arm_cntpct_time_source = {
 };
 
 REGISTER_TIME_SOURCE(arm_cntpct_time_source)
+
```
1 blank line is enough

lws-team commented:
Travis says,

```
readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 
readelf: Warning: Unknown ARM compact model index encountered
readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 
readelf: Warning: Unknown ARM compact model index encountered
readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 
readelf: Warning: Unknown ARM compact model index encountered
readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 
readelf: Warning: Unknown ARM compact model index encountered
arm-linux-gnueabihf-ld: OP-TEE can't fit init part into available physical memory
make: *** [out/arm-plat-vexpress/core/tee.elf] Error 1

The command "$make CFG_WITH_PAGER=y CFG_WITH_LPAE=y CFG_RPMB_FS=y CFG_SQL_FS=y CFG_DT=y CFG_PS2MOUSE=y CFG_PL050=y CFG_PL111=y CFG_TEE_CORE_LOG_LEVEL=1 CFG_TEE_CORE_DEBUG=y DEBUG=1" exited with 2.
```

Is this really related to my changes, or something else?  There is no big code and almost no allocations, just a few static chars and so.

lws-team commented:
CNTPCT seems to work just as well... so I just used that, thanks for the suggestion.  It obviates the need for the vcnt accessors... and the vertical whitespace is cleaned.

lws-team commented:
Yeah you're right.... there is a cycle counter but this isn't it.  It's only a thousand times better precision.  I fixed the comment and changed the patch headers.

jforissier commented:
Cycle counter... yes... but unfortunately not as easy to access as the CNTPCT. For the PMU PNCCNTR you need to know the base address as it's memory-mapped. I.e., platform-dependent...

I'm OK with your patches provided the Travis link issue is fixed of course. My guess would be you're just unlucky and have hit a size limit, maybe increasing `CFG_CORE_TZSRAM_EMUL_SIZE` for plat-vexpress would do the trick. @jenswi-linaro what do you say?

jenswi-linaro commented:
Yes, `CFG_CORE_TZSRAM_EMUL_SIZE` is likely the problem.

lws-team commented:
12.85s$ $make CFG_WITH_PAGER=y CFG_WITH_LPAE=y CFG_RPMB_FS=y CFG_SQL_FS=y CFG_DT=y CFG_PS2MOUSE=y CFG_PL050=y CFG_PL111=y CFG_TEE_CORE_LOG_LEVEL=1 CFG_TEE_CORE_DEBUG=y DEBUG=1

readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 

readelf: Warning: Unknown ARM compact model index encountered

readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 

readelf: Warning: Unknown ARM compact model index encountered

readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 

readelf: Warning: Unknown ARM compact model index encountered

readelf: Warning: Corrupt ARM compact model table entry: e12fff1e 

readelf: Warning: Unknown ARM compact model index encountered

arm-linux-gnueabihf-ld: Too few free pages to initialize paging

make: *** [out/arm-plat-vexpress/core/tee.elf] Error 1

ugh... try 360KiB

jforissier commented on core/tee/tee_cryp_utl.c:
```diff
@@ -378,6 +379,20 @@ TEE_Result tee_prng_add_entropy(const uint8_t *in, size_t len)
+ * platform-specific jitter entropy.  This implementation
+ * does not efficiently deliver entropy and is here for
+ * backwards-compatibility.
+ */
```
Please format this comment better:
- Sentences should start with an uppercase letter
- No double space
- Standard line width is 80 characters (applies to other comments as well)

jforissier commented on core/arch/arm/kernel/tee_time_arm_cntpct.c:
```diff
@@ -54,3 +56,40 @@ static const struct time_source arm_cntpct_time_source = {
+
+/*
+ * We collect jitter using cntpct in 32- or 64-bit
+ * mode that is clocked at the cpu core rate.
```
Comment still needs fixing.
In addition, abbreviations should be upper case (`s/cpu/CPU/` and `s/cntpct/CNTPCT/`).

lws-team commented:
One space between sentences... pfft.  My 1980-vintage RSA and Pitman's qualifications say different.  It's not even checkpatch arguing with it... still here you go.

jforissier commented on core/arch/arm/kernel/tee_time_arm_cntpct.c:
```diff
@@ -54,3 +56,39 @@ static const struct time_source arm_cntpct_time_source = {
+
+/*
+ * We collect jitter using CNTPCT in 32- or 64-bit mode that is clocked at the
+ * CPU core rate.
```
As I said twice already: `CNTPCT` is *not* clocked at the CPU core rate, it is clocked at a frequency of `CNTFRQ`.

Which raises another question now that I'm thinking about it. In low-power states, `CNTFRQ` may be lowered and `CNTPCT` incremented by more than 1 on each clock tick (to compensate for the lower update rate). See for instance the example given in the ARMv8 ARM (section G5.1.1 System counter). If the counter is incremented by steps of 500, you obviously won't be able to gather *any* entropy at all by sampling the two least significant bits.

lws-team replied:
You said it once, it's true.  But you just waved your hand at it last time in a way I was unable to interpret.

The "lumpy increment" is news to me but it makes sense.  Where the interval is too large that in fact nothing was added to the counter, nobody can do anything, and your old implementation did nothing useful either.

However when the counter has changed, we can get entropy from it by XORing every 2-bit pair in the counter on to the 2-bit slot being updated in turn.

I also changed the accumulator to rotate, since we don't need to reset the two bits with all those XORs either.

So thanks for the tip ^^

jforissier commented:
Out of curiosity, I have run the `dieharder` test suite on 3 data sets:
- 1 GB of data read from `/dev/urandom` on my PC
- 1 GB of data returned by `TEE_GenerateRandom()` 64 bytes at a time, without this PR (master)
- Same with this PR applied.

Results are as follows:
```
======= urandom
FAILED	0
WEAK	4
PASSED	110
======= rand_hikey_optee_928468c (master)
FAILED	1
WEAK	5
PASSED	108
======= rand_hikey_optee_0bee5af (this PR)
FAILED	0
WEAK	5
PASSED	109
```
More details here: https://pastebin.linaro.org/view/0e037049
So it seems this PR would improve things slightly, in any case it won't break the `dieharder` tests.

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org>`

lws-team commented:
Thanks.

Since the PRNG is basically the same, it's difficult to measure any change... we're adding a little trickle of noise *behind* the PRNG.  Logically though, it should indeed be a little better, particularly the initial seed.

jforissier commented:
@lws-team care to rebase and apply my R-b/T-b tags? Thanks.
